### PR TITLE
split SparsityPattern into a base class and derived class with a spec…

### PR DESCRIPTION
**Pull request by davydden**
…ial treatment for diagonal matrices

The idea is to have a base class which provides CSR and iterators/accessors without being limited to specific design choices of `SparsityPattern`, in particular storing diagonal elements first.

~~Possibly more functionality could be moved to the base class, the current PR is the minimum that I would need to use the base class and derive from it another one as shown in the unit test. All `tests/sparsity` pass for me locally.~~ EDTI: I moved as much as I could to the base class.

fixes https://github.com/dealii/dealii/issues/6917 

TODO:

- [x] adjust documentation

### CODE REVIEW

File: include/deal.II/lac/sparsity_pattern.h [hunk 1]
```diff
@@ -358,8 +867,14 @@ class SparsityPattern : public Subscriptor
    * Since the iterator does not allow to modify the sparsity pattern, this
    * type is the same as that for @p const_iterator.
    */
-  using iterator = SparsityPatternIterators::Iterator;
+  using iterator = SparsityPatternBase::iterator;
 
+  /**
+   * Since this class has to implement of the flavours of reinit() functions,
```
Comment (bangerth) [inline include/deal.II/lac/sparsity_pattern.h]:
```suggestion
   * Since this class has to implement some of the flavors of reinit() functions,
```
?

**Comment by bangerth** [comment]
The patch is hard to read because stuff is being moved around. Can you explain what you moved and what you didn't? Is it everything that has to do with providing memory that's now in the base class, and everything that interprets what is being stored and where that remains in the derived class?

What are the use cases you want to support? Is it just the issue of diagonal first?

**Comment by davydden** [comment]
>  Can you explain what you moved and what you didn't? Is it everything that has to do with providing memory that's now in the base class, and everything that interprets what is being stored and where that remains in the derived class?

More or less. Everything that does on use diagonal convention went to base class (iterators, accessors, etc). Whereas, all copy_from remained in derived class. Derived should implement one reinit() which is called from all other reinits.

> What are the use cases you want to support? Is it just the issue of diagonal first?

Personally I need static sparsity without diagonals-first.

**Comment by davydden** [comment]
> Can you explain what you moved and what you didn't?

it's probably easier to see that based on definitions of functions, both in `.h` and in `.cc`.

File: include/deal.II/lac/sparsity_pattern.h [hunk 2]
```diff
@@ -1398,6 +1567,39 @@ SparsityPattern::n_nonzero_elements() const
...
+}
+
+
+
+template <class Archive>
+inline void
+SparsityPatternBase::load(Archive &ar, const unsigned int)
+{
+  // forward to serialization function in the base class.
+  ar &static_cast<Subscriptor &>(*this);
```
Comment (davydden) [inline include/deal.II/lac/sparsity_pattern.h]: the annoying thing I don't understand is that I was not able to use the parent class `save/load` in the derived as
```
ar &static_cast<const SparsityPatternBase &>(*this);
ar &store_diagonal_first_in_row;
```

I would expect this to work without any issues, unless there is something related to `boost::serialization::make_array` that prevents serializing scalars after it...

Comment (davydden) [reply include/deal.II/lac/sparsity_pattern.h]: ok, interesting and wierd part is that base class serialization is never called on the line above:
```
ar &static_cast<const SparsityPatternBase &>(*this);
```

neither is it called if I try to do it the recommended way:
```
ar &boost::serialization::base_object<const SparsityPatternBase>(*this);
```

😲 

will keep looking...

Comment (davydden) [reply include/deal.II/lac/sparsity_pattern.h]: should be solved now.

**Comment by davydden** [comment]
@bangerth can we merge this or would you like me to do something more here?

**Comment by davydden** [comment]
> bangerth commented 7 days ago

@bangerth ping.

**Comment by davydden** [comment]
> bangerth commented 19 days ago

given that @bangerth commented on my other PR, perhaps I this is a lucky day and we can finally get through this one as well ;-)

**Comment by bangerth** [comment]
On 3/19/19 10:28 AM, Denis Davydov wrote:
> given that @bangerth <https://github.com/bangerth> commented on my other
> PR, perhaps I this is a lucky day and we can finally get through this one
> as well ;-)

Ugh. It's spring break, and I was about to go on a longish bike ride...

I'll put it on my list for tomorrow!

**Comment by bangerth** [comment]
OK from my side. There is one tester still pending, feel free to merge once that one is done.

**Comment by bangerth** [comment]
You also might want to squash the two commits, given that there is a bug in the version after the first commit alone.

**Comment by davydden** [comment]
> You also might want to squash the two commits, given that there is a bug in the version after the first commit alone.

will do.

**Pull request merged**